specs/opcode/35CALLDATALOAD.md

```diff
@@ -0,0 +1,40 @@
+# CALLDATALOAD opcode
+
+## Procedure
+
+The `CALLDATALOAD` opcode gets input data of current environment.
+
+## EVM Behaviour
+
+Stack input is the byte offset to read call data from. Stack output is a 32-byte value starting from the given offset of the call data. All bytes after the end of the call data are set to `0`.
+
+## Constraints
+
+1. opId == 0x35
+2. State Transition:
+   - if is_root_call:
+     - rw_counter += 3 (1 stack read, 1 call context read, 1 stack write)
+   - if is_internal_call:
+     - rw_counter += rw_counter_offset ∈ {4, 5, ..., 35, 36} (1 stack read, 2 call context reads, i ∈ {0, 1, ..., 31, 32} memory reads, 1 stack write)
+   - stack_pointer unchanged
+   - pc + 1
+   - gas - 3
+3. Lookups:
+   - `offset` is at the top of the stack
+   - `tx_id` is in the RW table (call context)
+   - if is_root_call (where `src_addr = offset`):
+     - `calldata_length` is in the TX table
+     - i ∈ {0, 1, ..., 31, 32} lookups for `i in range(32)`: if `buffer.read_flag(i)` then the i'th byte of the element on top of the stack `calldata_word[i]` is in the TX table {tx id, call data, src_addr + i}
```

Comment:
should be i ∈ {0, 1, ..., 31} ?

Reply:
For the case where no byte is actually read (`out of bound` or `calldatalength=0`, etc.) we have `0` lookups for from the tx table / call context. But if we read 32 bytes, there are `32` lookups. `i` is not the index of the byte being looked up, but the number of lookups, hence it belongs to a set of `{0, 1, 2, ..., 31, 32}` here.

Reply:
Ohhh this was num of lookups.

Sorry, I saw `i` used later as the range index and I thought that it was not meaning num of lookups.